Support for defining DNF mirrors in builder.conf: typically local mirrors

fepitre opened the pull request:
Here is what I propose as an alternative of using REPO_PROXY: we need to define in the builder.conf the variable REPO_BASEURL_PREFIX. Then, the main and update repos are defined as:

$(REPO_BASEURL_PREFIX)/fedora/linux/releases/$(subst fc,,$(DIST))/Everything and $(REPO_BASEURL_PREFIX)/fedora/linux/updates/$(subst fc,,$(DIST))/x86_64/.

So for example one can define REPO_BASEURL_PREFIX=http://mylocalmirror/pub following the default structure of Fedora mirrors.

It uses "baseurl" but the problem is that it does not override the metalink option and it seems there is not any cmd line option to delete the option. In consequence, we have to remove it manually when constructing the chroot env.

marmarek commented:
In builder-debian we already have similar thing: DEBIAN_MIRRORS. Can you rename it to FEDORA_MIRRORS or sth like this?

fepitre commented:
No problem. I will also modify it in linux-template-builder as you merged my PR.

marmarek commented on prepare-chroot-base:
```diff
@@ -56,9 +56,20 @@ if ! [ -f "${INSTALLDIR}/tmp/.prepared_base" ]; then
+    # in the repo file because baseurl seems to not override the metalink
+    if [ "x${FEDORA_MIRRORS}" != "x" ]; then
+        sed -e "s/\\\$releasever/${DIST#fc}/g" \
+            -e "/metalink/d" \
```
Better comment out instead of removing. And match with `^metalink` to not comment out already commented line.

marmarek commented on prepare-chroot-base:
```diff
@@ -93,5 +104,11 @@ if ! [ -f "${INSTALLDIR}/tmp/.prepared_base" ]; then
 
+    # TODO: check if there is a way to force baseurl to override metalink
+    if [ "x$FEDORA_MIRRORS" != "x" ]; then
+        sed -i "/metalink/d" "${INSTALLDIR}/etc/yum.repos.d/fedora.repo" "${INSTALLDIR}/etc/yum.repos.d/fedora-updates.repo"
```
And here too.

marmarek commented on prepare-chroot-base:
```diff
@@ -93,5 +104,11 @@ if ! [ -f "${INSTALLDIR}/tmp/.prepared_base" ]; then
         cp -a /dev/$f $INSTALLDIR/dev/
     done
 
+    # TODO: check if there is a way to force baseurl to override metalink
```
Well, you can create own metalink pointing to your local mirror... It's just an xml file. But probably not worth it.

fepitre replied:
Indeed, I also though about a .repo file directly but it means to create for each FC a repo/xml file...

fepitre replied:
We could imagine to create in the future something doing it automatically.

marmarek commented on template_scripts/distribution.sh:
```diff
@@ -10,6 +10,11 @@ if [ -n "${REPO_PROXY}" ]; then
 fi
 
+if [ -n "${FEDORA_MIRRORS}" ]; then
+    YUM_OPTS="$YUM_OPTS --setopt=fedora.baseurl=${FEDORA_MIRRORS}/fedora/linux/releases/${DIST/fc/}/Everything/x86_64/os/"
```
Oh, but this actually expect just one mirror, not a list of them. So, either get just first entry (word) of FEDORA_MIRRORS (and care about supporting multiple of them later, if at all). Or rename to FEDORA_MIRROR...
Maybe just add something like this at the top:
```
FEDORA_MIRROR=$(echo "$FEDORA_MIRRORS" | awk '{print $1}')
```

fepitre replied:
Supporting multiple mirrors would it be necessary/useful? Indeed, we can let this possibility and do as you suggest.

marmarek replied:
I wouldn't implement multiple mirrors right now. But it could be useful to have some fallback automatically. For example the most common reason by build fails on travis if failing to synchronize repository metadata...

fepitre replied:
Ok. So I rename everything to FEDORA_MIRROR. The best next step would be to generate a metalink xml file from a list of URLS. Do you agree?

marmarek replied:
👍

fepitre commented:
It should be good. This has to be completed with the PR you approved 'DNF' as $YUM_OPTS is needed when building the chroot.

fepitre commented:
I put things related to xinit in gui-agent-linux (https://github.com/fepitre/qubes-gui-agent-linux/commit/53c2781d3bfae2925c5a6bb5738a45653b51b859). I also add Thunar actions. it generates gui-agent-xfce rpm. In consequence, I added in 04_install_qubes script the flavor selection with respect to a qubes packages list. For XFCE, I used the alread packages_qubes.list where I added gui-agent-xfce. The template is built but I can only test it tonight or tomorrow (not at home currently).
